### ISSUE TITLE
Permission: SQL questions cannot be viewed after granted "View access"

### ISSUE THREAD

**Issue by vincentimo** [metabase/metabase#11672]
**Describe the bug**

There's a permission bug in Metabase 0.34.0 I'd like to report.

According to [Metabase Documentation about Collections](https://www.metabase.com/docs/latest/administration-guide/06-collections):

> **View access:**  the user can see all the questions, dashboards, and pulses in the collection. If the user does not have permission to view some or all of the questions included in a given dashboard or pulse then those questions will not be visible to them; but any questions that are saved in this collection  *will*  be visible to them,  *even if the user doesn’t have access to the underlying data used to in the question.*

This works as expected for Simple/Custom questions, but does not work as expected for SQL questions. In other words, SQL questions cannot be viewed by a user even after that user is granted either `View access` or `Curate access` to the collection where the question is located, and the Data Permission is either `No Access` or `Data Access`.

It has two different outputs depending on the Data Permission:

* Case 1: If Data Permission is `No Access`, then there is black pair of keys figure with black text saying "Sorry, you don’t have permission to see that." (Figure 1)
* Case 2: If Data Permission is `Data Access`, then there is red diagonal cross inside circle figure with red text saying "You do not have permissions to run this query." (Figure 2)

Thank you for your help.

**Logs**

For case 1:

```
[38f155d6-c8ab-47d9-83c6-6e0a4d33147f] 2020-01-10T10:08:24+07:00 WARN metabase.middleware.log POST /api/dataset 403 3.6 ms (3 DB calls)
"You don't have permissions to do that."
```

For case 2:

```
[38f155d6-c8ab-47d9-83c6-6e0a4d33147f] 2020-01-10T10:08:59+07:00 WARN metabase.query-processor.middleware.process-userland-query Query failure {:status :failed,
:class clojure.lang.ExceptionInfo,
:error "You do not have permissions to run this query.",
:stacktrace
("--> query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:23)"
"query_processor.middleware.permissions$perms_exception.invoke(permissions.clj:22)"
"query_processor.middleware.permissions$fn__42264$check_ad_hoc_query_perms__42269$fn__42270.invoke(permissions.clj:32)"
"query_processor.middleware.permissions$fn__42264$check_ad_hoc_query_perms__42269.invoke(permissions.clj:28)"
"query_processor.middleware.permissions$fn__42287$check_query_permissions_STAR___42292$fn__42298.invoke(permissions.clj:40)"
"query_processor.middleware.permissions$fn__42287$check_query_permissions_STAR___42292.invoke(permissions.clj:34)"
"query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__40540.invoke(cumulative_aggregations.clj:57)"
"query_processor.middleware.resolve_joins$resolve_joins$fn__42873.invoke(resolve_joins.clj:184)"
"query_processor.middleware.limit$limit$fn__41651.invoke(limit.clj:19)"
"query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__45642.invoke(results_metadata.clj:87)"
"query_processor.middleware.format_rows$format_rows$fn__41639.invoke(format_rows.clj:76)"
"query_processor.middleware.add_dimension_projections$add_remapping$fn__37181.invoke(add_dimension_projections.clj:232)"
"query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__37904.invoke(add_source_metadata.clj:107)"
"query_processor.middleware.resolve_source_table$resolve_source_tables$fn__42923.invoke(resolve_source_table.clj:46)"
"query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__37741.invoke(add_row_count_and_status.clj:16)"
"query_processor.middleware.driver_specific$process_query_in_context$fn__40615.invoke(driver_specific.clj:12)"
"query_processor.middleware.resolve_driver$resolve_driver$fn__42537.invoke(resolve_driver.clj:22)"
"query_processor.middleware.store$initialize_store$fn__45667$fn__45668.invoke(store.clj:11)"
"query_processor.store$do_with_store.invokeStatic(store.clj:46)"
"query_processor.store$do_with_store.invoke(store.clj:40)"
"query_processor.middleware.store$initialize_store$fn__45667.invoke(store.clj:10)"
"query_processor.middleware.async$async__GT_sync$fn__22968.invoke(async.clj:23)"
"query_processor.middleware.async_wait$runnable$fn__39657.invoke(async_wait.clj:89)"),
:query
{:native {:query "select * from ORDERS;", :template-tags {}},
:type "native",
:parameters [],
:async? true,
:middleware {:add-default-userland-constraints? true, :userland-query? true},
:info
{:executed-by 2,
:context :ad-hoc,
:card-id nil,
:nested? false,
:query-hash [66, -81, -62, 127, -9, -7, 30, -76, -35, -73, -94, 12, 29, -48, 3, 40, 103, -60, -102, 88, -19, 93, 126, 10, -98, -97, 5, 83, 88, 18, -22, 94]},
:constraints {:max-results 10000, :max-results-bare-rows 2000}},
:ex-data {:required-permissions #{"/db/1/native/"}, :actual-permissions #{"/collection/2/" "/collection/root/" "/db/1/schema/"}, :permissions-error? true}}
```

**To Reproduce**

  1. Run metabase.jar for the first time
  2. Create admin account, e.g. `admin@example.com`
  3. As Admin, in People menu, create member account, e.g. `member@example.com`
  4. As Admin, in People menu, create group, e.g. `Playground`
  5. As Admin, in People menu, add `member@example.com` to `Playground`
  6. As Admin, in Permissions menu, adjust access for `All Users` for `Sample Dataset` to be either `No access` or `Data Access`, depending on we want to reproduce result in Figure 1 or Figure 2
  7. As Admin, in Permissions menu, grant `Playground` for `Our analytics` to be either `View collection` or `Curate collection` (does not matter)
  8. As Admin, create a question using native query, e.g. `select * from ORDERS;`
  9. As Admin, save the question in `Our analytics`, e.g. as `SQL question`
  10. As Member, try to view the question
  11. Member would not be able to view the question, and the output is either Figure 1 or Figure 2

**Expected behavior**

Metabase displays data in a question located in a particular collection, when a user is granted either `View access` or `Curate access` to that collection, regardless of Simple/Custom questions or SQL questions.

**Screenshots**

[**Figure 1** Case 1: The output if Data Permission is `No Access`](https://discourse.metabase.com/uploads/default/original/2X/c/ce185e979bf9360d6561f4728e38fe99923ecb3e.png)

[**Figure 2** Case 2: The output if Data Permission is `Data Access`](https://discourse.metabase.com/uploads/default/original/2X/b/b1a721aef0d137d69178b22d8d2696412216bb58.png)

**Information about your Metabase Installation**

You can get this information by going to Admin -> Troubleshooting.

- Your browser and the version: Chrome Version 79.0.3945.88 (Official Build) (64-bit)
- Your operating system: macOS Mojave Version 10.14.6
- Your databases: BigQuery
- Metabase version: 0.34.0
- Metabase hosting environment: Docker
- Metabase internal database: PostgreSQL

**Severity**

This issue blocks some users and we have to manually grant the users' Data Permission to `SQL Queries`. The users could view the data not intended for them.

**Additional context**

This bug has also been reported in https://discourse.metabase.com/t/permission-sql-questions-cannot-be-viewed-after-granted-view-access/8533.

**Comment by mazameli** [comment]
We believe this issue was resolved by https://github.com/metabase/metabase/pull/11579. This fix will be in 0.34.1.

**Comment by camsaul** [comment]
resolved by #11579
